File: flutter/test/domain/algorithms/shift_metrics_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:scheduler_cp/domain/models/shift_type.dart';
import 'package:scheduler_cp/domain/models/shift_cycle_config.dart';
import 'package:scheduler_cp/domain/algorithms/shift_calculator.dart';
import 'package:scheduler_cp/domain/algorithms/shift_metrics.dart';

void main() {
  final refDate = ShiftCycleConfig.referenceDate; // 2025-12-15

  group('countShiftTypeInMonth', () {
    test('returns non-negative for all shift types', () {
      for (final type in ShiftType.values) {
        final count =
            countShiftTypeInMonth(2026, 5, type, referenceDate: refDate);
        expect(count, isNonNegative);
      }
    });

    test('total counts = days in month', () {
      int total = 0;
      for (final type in ShiftType.values) {
        total +=
            countShiftTypeInMonth(2026, 5, type, referenceDate: refDate);
      }
      expect(total, 31);
    });
  });

  group('countWorkDaysInMonth', () {
    test('work days + rest + study = month total', () {
      final workDays =
          countWorkDaysInMonth(2026, 5, referenceDate: refDate);
      final restDays = countShiftTypeInMonth(2026, 5, ShiftType.REST,
          referenceDate: refDate);
      final studyDays = countShiftTypeInMonth(2026, 5, ShiftType.STUDY,
          referenceDate: refDate);
      expect(workDays + restDays + studyDays, 31);
    });

    test('work days count is within month bounds', () {
      final count = countWorkDaysInMonth(2026, 5, referenceDate: refDate);
      expect(count, isNonNegative);
      expect(count, lessThanOrEqualTo(31));
    });
  });

  group('consecutiveWorkDays', () {
    test('returns 0 if today is REST', () {
      var date = refDate;
      while (getShiftTypeForDate(date, referenceDate: refDate) !=
          ShiftType.REST) {
        date = date.add(const Duration(days: 1));
      }
      expect(
          consecutiveWorkDays(date, referenceDate: refDate), 0);
    });

    test('returns >0 if today is work day', () {
      final count =
          consecutiveWorkDays(refDate, referenceDate: refDate);
      expect(count, greaterThan(0));
    });
  });

  group('daysUntilNextRest', () {
    test('returns within cycle length bounds', () {
      final days = daysUntilNextRest(refDate, referenceDate: refDate);
      expect(days, isNonNegative);
      expect(days, lessThanOrEqualTo(43));
    });

    test('returns 0 when next day is REST', () {
      var date = refDate;
      while (getShiftTypeForDate(date.add(const Duration(days: 1)),
              referenceDate: refDate) !=
          ShiftType.REST) {
        date = date.add(const Duration(days: 1));
      }
      expect(daysUntilNextRest(date, referenceDate: refDate), 0);
    });
  });
}

File: android/app/src/test/java/com/simpleshift/scheduler/domain/ShiftMetricsTest.kt
package com.simpleshift.scheduler.domain

import com.simpleshift.scheduler.domain.model.ShiftType
import org.junit.Assert.assertEquals
import org.junit.Test
import java.time.LocalDate
import java.time.YearMonth

class ShiftMetricsTest {

    private val twoDayCycle = listOf(ShiftType.MORNING, ShiftType.REST)

    @Test
    fun `countShiftTypeInMonth 2-day cycle Feb 2026 REST`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(14, countShiftTypeInMonth(ym, ShiftType.REST, customCycle = twoDayCycle))
    }

    @Test
    fun `countShiftTypeInMonth 2-day cycle Feb 2026 MORNING`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(14, countShiftTypeInMonth(ym, ShiftType.MORNING, customCycle = twoDayCycle))
    }

    @Test
    fun `countWorkDaysInMonth 2-day cycle Feb 2026`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(14, countWorkDaysInMonth(ym, customCycle = twoDayCycle))
    }

    @Test
    fun `consecutiveWorkDays backtracks correctly`() {
        val today = LocalDate.of(2025, 12, 15) // offset 0 = MORNING
        assertEquals(1, consecutiveWorkDays(today, customCycle = twoDayCycle))
    }

    @Test
    fun `consecutiveWorkDays returns 0 when today is REST`() {
        val today = LocalDate.of(2025, 12, 16) // offset 1 = REST
        assertEquals(0, consecutiveWorkDays(today, customCycle = twoDayCycle))
    }

    @Test
    fun `consecutiveWorkDays returns 0 when today is STUDY`() {
        val studyCycle = listOf(ShiftType.STUDY, ShiftType.MORNING)
        val today = LocalDate.of(2025, 12, 15)
        assertEquals(0, consecutiveWorkDays(today, customCycle = studyCycle))
    }

    @Test
    fun `daysUntilNextRest returns 0 when tomorrow is REST`() {
        val today = LocalDate.of(2025, 12, 15) // MORNING, tomorrow REST
        assertEquals(0, daysUntilNextRest(today, customCycle = twoDayCycle))
    }

    @Test
    fun `daysUntilNextRest when today is REST searches from tomorrow`() {
        val today = LocalDate.of(2025, 12, 16) // REST, tomorrow MORNING, next REST the day after
        assertEquals(1, daysUntilNextRest(today, customCycle = twoDayCycle))
    }

    @Test
    fun `daysUntilNextRest returns 0 when no REST in cycle`() {
        val allWork = listOf(ShiftType.MORNING)
        val today = LocalDate.of(2025, 12, 15)
        assertEquals(0, daysUntilNextRest(today, customCycle = allWork))
    }

    @Test
    fun `countWorkDaysInMonth all work returns full month`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(28, countWorkDaysInMonth(ym, customCycle = listOf(ShiftType.MORNING)))
    }

    @Test
    fun `countWorkDaysInMonth all REST returns 0`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(0, countWorkDaysInMonth(ym, customCycle = listOf(ShiftType.REST)))
    }

    @Test
    fun `countWorkDaysInMonth STUDY is not work`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(0, countWorkDaysInMonth(ym, customCycle = listOf(ShiftType.STUDY)))
    }

    @Test
    fun `countWorkDaysInMonth NIGHT is work`() {
        val ym = YearMonth.of(2026, 2)
        assertEquals(28, countWorkDaysInMonth(ym, customCycle = listOf(ShiftType.NIGHT)))
    }

    @Test
    fun `countShiftTypeInMonth with default cycle returns valid count`() {
        val ym = YearMonth.of(2026, 2)
        // Just verify it doesn't crash and returns something in valid range
        val count = countShiftTypeInMonth(ym, ShiftType.REST)
        assert(count in 0..28) { "REST count should be 0..28, was $count" }
    }

    @Test
    fun `countWorkDaysInMonth with default cycle returns valid count`() {
        val ym = YearMonth.of(2026, 2)
        val count = countWorkDaysInMonth(ym)
        assert(count in 0..28) { "Work days should be 0..28, was $count" }
    }
}
